src/passes/Outlining.cpp

```diff
@@ -13,23 +13,225 @@
+    instrCounter = 0;
+    seqCounter = 0;
+    state = NotInSeq;
+    DBG(std::cout << "\n\n$" << func->name << " Func Start "
+                  << &existingBuilder);
+  }
+  ReconstructState getCurrState() {
+    // We are either in a sequence or not in a sequence. If we are in a sequence
+    // and have already created the body of the outlined function that will be
+    // called, then we will skip instructions, otherwise we add the instructions
+    // to the outlined function. If we are not in a sequence, then the
+    // instructions are sent to the existing function.
+    if (seqCounter < sequences.size() &&
+        instrCounter >= sequences[seqCounter].startIdx &&
+        instrCounter < sequences[seqCounter].endIdx) {
+      return getModule()->getFunction(sequences[seqCounter].func)->body
+               ? InSkipSeq
+               : InSeq;
+    }
+    return NotInSeq;
+  }
+  void maybeBeginSeq() {
+    instrCounter++;
+    auto currState = getCurrState();
+    if (currState != state) {
+      switch (currState) {
+        case NotInSeq:
+          return;
+        case InSeq:
+          transitionToInSeq();
+          break;
+        case InSkipSeq:
+          transitionToInSkipSeq();
+          break;
+      }
+    }
+    state = currState;
+  }
+  void transitionToInSeq() {
+    Function* outlinedFunc =
+      getModule()->getFunction(sequences[seqCounter].func);
+    ASSERT_ERR(outlinedBuilder.visitFunctionStart(outlinedFunc));
+
+    // Add a local.get instruction for every parameter of the outlined function.
+    Signature sig = outlinedFunc->type.getSignature();
+    for (Index i = 0; i < sig.params.size(); i++) {
+      ASSERT_ERR(outlinedBuilder.makeLocalGet(i));
+    }
+
+    // Make a call from the existing function to the outlined function. This
+    // call will replace the instructions moved to the outlined function
+    ASSERT_ERR(existingBuilder.makeCall(outlinedFunc->name, false));
+    DBG(std::cout << "\ncreated outlined fn: " << outlinedFunc->name);
+  }
+
+  void transitionToInSkipSeq() {
+    Function* outlinedFunc =
+      getModule()->getFunction(sequences[seqCounter].func);
+    ASSERT_ERR(existingBuilder.makeCall(outlinedFunc->name, false));
+    DBG(std::cout << "\n\nstarting to skip instructions "
+                  << sequences[seqCounter].startIdx << " - "
+                  << sequences[seqCounter].endIdx - 1 << " to "
+                  << sequences[seqCounter].func
+                  << " and adding call() instead");
+  }
+  void maybeEndSeq() {
+    if (instrCounter + 1 == sequences[seqCounter].endIdx) {
+      transitionToNotInSeq();
+      state = NotInSeq;
+    }
+  }
+  void transitionToNotInSeq() {
+    if (state == InSeq) {
+      ASSERT_ERR(outlinedBuilder.visitEnd());
+      DBG(std::cout << "\n\nEnd of sequence to " << &outlinedBuilder);
+    }
+    // Completed a sequence so increase the seqCounter and reset the state
+    seqCounter++;
+  }
+
+#define RECONSTRUCT_DEBUG 0
```

Comment:
It doesn't look like this is used anywhere any more.

Reply:
done